### PR TITLE
build: impl serialization for storage

### PR DESCRIPTION
**Pull request by AvivYossef-starkware**
Introduces storage_serialize_test function to generate and serialize a MapStorage instance containing StorageKey and StorageValue pairs for values in the range 0..=99. Enabling serialization of MapStorage to JSON format using Serde.

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)

### CODE REVIEW

File: crates/committer/src/storage/storage_trait.rs
Comment (AvivYossef-starkware) [inline crates/committer/src/storage/storage_trait.rs:58]: Its needed since serde couldn't serialize hashmap if key != String

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 2 of 7 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)* status: all files reviewed, 5 unresolved discussions (waiting on @amosStarkware and @AvivYossef-starkware)

___
*[`crates/committer/src/storage/storage_trait.rs` line 58 at r1](https://reviewable.io/reviews/starkware-libs/committer/82#gh-1586491907:-Nx6rz3OCx9WigVKrZhz:bkc2yw2) ([raw file](https://github.com/starkware-libs/committer/blob/73c145cbf34830926d34072a116817d1ac03d57d/crates/committer/src/storage/storage_trait.rs#L58)):*
<details><summary><i>Previously, AvivYossef-starkware wrote…</i></summary><blockquote>

Its needed since serde couldn't serialize hashmap if key != String 
</blockquote></details>

this means deserializing side will need to convert hex to bytes, right?
___
*[`crates/committer/src/storage/storage_trait.rs` line 70 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6rahl1WvjwTM3N2ZU:-Nx6rahl1WvjwTM3N2ZV:bns4yoe) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer/src/storage/storage_trait.rs#L70)):*
> ```XML
>         S: Serializer,
>     {
>         // Convert Vec<u8> to hexadecimal string representation
> ```

punctuation, grammar, capitalization



_Suggestion:_
```XML
    /// Serializes `StorageKey` to hexadecimal string representation.
    /// Needed since serde's Serialize derive attribute only works on
    /// HashMaps with String keys.
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert Vec<u8> to hexadecimal string representation.
```
___
*[`crates/committer/src/storage/storage_trait.rs` line 72 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6rrsI6oUoLVQdCoRI:-Nx6rrsI6oUoLVQdCoRJ:bwyqodn) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer/src/storage/storage_trait.rs#L72)):*
> ```XML
>         // Convert Vec<u8> to hexadecimal string representation
>         let hex_string = hex::encode(&self.0);
>         serializer.serialize_str(&hex_string)
> ```

redundant variable? or is the value dropped?



_Suggestion:_
```XML
        serializer.serialize_str(&hex::encode(&self.0))
```
___
*[`crates/committer_cli/src/tests/python_tests.rs` line 414 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sagvEPxLCNVav07N:-Nx6sagvEPxLCNVav07O:btoht5h) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L414)):*
> ```XML
>         let value = StorageValue(Felt::from(i).as_bytes().to_vec());
>         storage.set(key, value);
>     }
> ```

please add test cases with actual serialized nodes and leaves; and on the deserializing side, check that the deserialized key matches the expected node/leaf keys



_Code quote:_
```XML
    for i in 0..=99_u128 {
        let key = StorageKey(Felt::from(i).as_bytes().to_vec());
        let value = StorageValue(Felt::from(i).as_bytes().to_vec());
        storage.set(key, value);
    }
```
___
*[`crates/committer_cli/src/tests/python_tests.rs` line 417 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sOkh0DERUynjEV1k:-Nx6sOki8O3JqlG9uwr_:bslkync) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L417)):*
> ```XML
> 
>     serde_json::to_string(&storage)
>         .unwrap_or_else(|error| panic!("Failed to serialize storage: {}", error))
> ```

the `PythonTestError` enum has a variant with `#[from] serde_json::Error`, why not use it?



_Suggestion:_
```XML
    serde_json::to_string(&storage).map_err(PythonTestError::from)
```

**Review by AvivYossef-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)* status: 5 of 7 files reviewed, 5 unresolved discussions (waiting on @amosStarkware and @dorimedini-starkware)

___
*[`crates/committer/src/storage/storage_trait.rs` line 58 at r1](https://reviewable.io/reviews/starkware-libs/committer/82#gh-1586491907:-Nx6tAU49TWISjGesyvl:bbw4lxl) ([raw file](https://github.com/starkware-libs/committer/blob/73c145cbf34830926d34072a116817d1ac03d57d/crates/committer/src/storage/storage_trait.rs#L58)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

this means deserializing side will need to convert hex to bytes, right?
</blockquote></details>

Right. 
___
*[`crates/committer/src/storage/storage_trait.rs` line 70 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6rahl1WvjwTM3N2ZU:-Nx6tqiK8jQBnpVa_blO:b-6aqlyu) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer/src/storage/storage_trait.rs#L70)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

punctuation, grammar, capitalization
</blockquote></details>

Sorry :|
___
*[`crates/committer/src/storage/storage_trait.rs` line 72 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6rrsI6oUoLVQdCoRI:-Nx6uuyg6AVWFW0C5SXN:b-4sl6g1) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer/src/storage/storage_trait.rs#L72)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

redundant variable? or is the value dropped?
</blockquote></details>

Redundant. But I thought it made the code cleaner
___
*[`crates/committer_cli/src/tests/python_tests.rs` line 414 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sagvEPxLCNVav07N:-Nx6v9pjCWoSNCpp09ip:bst7yyq) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L414)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

please add test cases with actual serialized nodes and leaves; and on the deserializing side, check that the deserialized key matches the expected node/leaf keys
</blockquote></details>

I'll add. I want to do it in different PR
___
*[`crates/committer_cli/src/tests/python_tests.rs` line 417 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sOkh0DERUynjEV1k:-Nx6wP-b2JKhRJqwXypy:b-896fix) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L417)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

the `PythonTestError` enum has a variant with `#[from] serde_json::Error`, why not use it?
</blockquote></details>

Done.

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)* status: all files reviewed, 1 unresolved discussion (waiting on @amosStarkware and @AvivYossef-starkware)

___
*[`crates/committer/src/storage/storage_trait.rs` line 72 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6rrsI6oUoLVQdCoRI:-Nx6xZ420QfxMKI5WP7L:b4hxigr) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer/src/storage/storage_trait.rs#L72)):*
<details><summary><i>Previously, AvivYossef-starkware wrote…</i></summary><blockquote>

Redundant. But I thought it made the code cleaner
</blockquote></details>

I am against used-once variables that don't decrease the number of lines, `hex::encode(self)` is very readable :)
___
*[`crates/committer_cli/src/tests/python_tests.rs` line 414 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sagvEPxLCNVav07N:-Nx6xnG33PEbu99HG9wA:b-9k7of1) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L414)):*
<details><summary><i>Previously, AvivYossef-starkware wrote…</i></summary><blockquote>

I'll add. I want to do it in different PR
</blockquote></details>

ok, please open the PR and I'll unblock (otherwise I'll forget)

**Review by AvivYossef-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)* status: all files reviewed, 1 unresolved discussion (waiting on @amosStarkware and @dorimedini-starkware)

___
*[`crates/committer/src/storage/storage_trait.rs` line 72 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6rrsI6oUoLVQdCoRI:-Nx6z3mj-8XN2foRFV6F:bwygxsx) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer/src/storage/storage_trait.rs#L72)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I am against used-once variables that don't decrease the number of lines, `hex::encode(self)` is very readable :)
</blockquote></details>

Noted.
___
*[`crates/committer_cli/src/tests/python_tests.rs` line 414 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sagvEPxLCNVav07N:-Nx6z7897sm5nzE2L7m4:b-3mm7xs) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L414)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

ok, please open the PR and I'll unblock (otherwise I'll forget)
</blockquote></details>

Ok, opened.

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)* status: all files reviewed, 1 unresolved discussion (waiting on @amosStarkware and @AvivYossef-starkware)

___
*[`crates/committer_cli/src/tests/python_tests.rs` line 414 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sagvEPxLCNVav07N:-Nx71e1iACmI8RP2R0jU:b-6flcli) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L414)):*
<details><summary><i>Previously, AvivYossef-starkware wrote…</i></summary><blockquote>

Ok, opened.
</blockquote></details>

paste a link here?

**Review by AvivYossef-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82)* status: all files reviewed, 1 unresolved discussion (waiting on @amosStarkware and @dorimedini-starkware)

___
*[`crates/committer_cli/src/tests/python_tests.rs` line 414 at r2](https://reviewable.io/reviews/starkware-libs/committer/82#-Nx6sagvEPxLCNVav07N:-Nx7210c9-GNlEtXj-_m:b-api0nr) ([raw file](https://github.com/starkware-libs/committer/blob/37175b2d99cfa50a332532bc72e3e7e5523bf2c1/crates/committer_cli/src/tests/python_tests.rs#L414)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

paste a link here?
</blockquote></details>

https://github.com/starkware-libs/committer/pull/89/files

**Review by dorimedini-starkware — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/82#-:-Nx73bR17EI7YSdJs6kK:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @amosStarkware)

**Pull request merged**
